src/wallet/node_rpc_proxy.cpp

```diff
@@ -217,4 +221,90 @@ boost::optional<std::string> NodeRPCProxy::get_fee_quantization_mask(uint64_t &f
   return boost::optional<std::string>();
 }
 
+const std::vector<cryptonote::COMMAND_RPC_GET_SERVICE_NODES::response::entry> *NodeRPCProxy::get_service_nodes(std::vector<std::string> const &pubkeys, boost::optional<std::string> &failed) const
```

Comment:
`shared_ptr` does not help with data races though. If expensive to copy, we can leave the reference maybe adding a comment that it can be changed by other thread. We can probably copy for a single node.

Reply:
Hmm that was sloppy of me. Leaving as a ref/ptr seems likely to break eventually, i.e. show_transfers on large wallets and blocks get mined interim. The data race is around the assignment to the cached version, I can move the mutex to around it, shared ptr was for perf on the caller's site, but that doesn't really make much sense.

After thinking abit, shall I just make this a copy? Keep it simple until it becomes an issue. Onus on caller to make it shared if they need be. The other would be storing a shared_ptr of the cached list and passing out that shared ptr around.

Reply:
I agree, copying shouldn't be a problem.

Reply:
Fixed.